### PR TITLE
Dockerfile

### PR DESCRIPTION
**Pull request by neosbob**
Pull request for all changes made to the "dockerfile" branch.

### CODE REVIEW

File: webserver.cc [hunk 1]
```diff
@@ -49,7 +49,8 @@ Server::Server(configArguments configArgs, std::map<std::string, std::vector<std
     signals.async_wait(boost::bind(&Server::stop, this));
 
     ip::tcp::resolver resolver(io_service);
-    ip::tcp::resolver::query query(IPADDR, std::to_string(configArgs.port));
+    ip::tcp::resolver::query query("0.0.0.0", std::to_string(configArgs.port));
```
Comment (yanzunh) [inline webserver.cc:52]: it might be worthwhile to put the ip address in the config file, so that we can change it easily

File: Makefile [hunk 2]
```diff
@@ -4,7 +4,18 @@ CXXFLAGS= -g -Wall -pthread -std=c++11 $(CXXOPTIMIZE)
 GTEST_DIR=googletest/googletest
 SERVERCLASSES=config_parser.cc
 
-default:  config_parser config_parser_test request_handler.o echo_handler.o file_handler.o not_found_handler.o status_handler.o request.o response.o mime_types.o session.o  echo_handler_test file_handler_test not_found_handler_test response_test request_test session_test http_client_test proxy_handler_test webserver
+default:  config_parser config_parser_test request_handler.o echo_handler.o file_handler.o not_found_handler.o status_handler.o request.o response.o mime_types.o session.o webserver
```
Comment (yanzunh) [inline Makefile]: it might be better to move config_parse_test to test as well

File: Makefile [hunk 3]
```diff
@@ -99,7 +108,7 @@ proxy_handler_test:
 clean:
 	rm -rf *.o *.a *~ *.gch *.swp *.dSYM *.gcda *.gcno *.gcov config_parser config_parser_test *.tar.gz webserver request_handler.o session.o request_handler_test webserver_test session_test echo_handler_test file_handler_test response_test request_test not_found_handler_test session_test http_client_test proxy_handler_test
 
-test: clean default
+test: clean echo_handler_test file_handler_test not_found_handler_test response_test request_test session_test http_client_test proxy_handler_test
```
Comment (echen94) [inline Makefile]: It's good that we separate out the tests. Might make it faster to compile with the default make.

**Review by echen94 — COMMENTED** [review]
Good implementation of Dockerfile. Good change to Makefile. I agree with Bobby about making config test out of default make and separating IP address to a place that is easy to configure.

File: webserver.cc (hunk 1 above)
Comment (neosbob) [reply webserver.cc:52]: Unless it is 0.0.0.0, the server will not accept external connections though.

File: Makefile (hunk 2 above)
Comment (neosbob) [reply Makefile]: good idea!

File: webserver.cc (hunk 1 above)
Comment (yanzunh) [reply webserver.cc:52]: cool!

**Review by neosbob — COMMENTED** [review]
Moved config_parser_test to the test make target. Done to isolate making tests and making webserver.

**Review by yanzunh — APPROVED** [review]
Very nice!

File: README.md [hunk 4]
```diff
@@ -195,3 +195,23 @@ In chrome browser:
 `localhost::<port>/<uri_prefix>/<file_name>`
 
 `<file_name>` is optional. It depends on the type of handler. 
+
+## Deployment
+
+`make build` to build and run webserver in docker
```
Comment (neosbob) [inline README.md:201]: Make build actually builds the large image file into a tar file using Dockerfile, then it saves that tar file as webserver.tar.

File: README.md [hunk 5]
```diff
@@ -195,3 +195,23 @@ In chrome browser:
...
+
+`make deploy` to  automate the deployment
+
+This `deploy` target builds the shrunken image, copies the image to the AWS-ec2, and runs make using the new image on AWS-ec2 server.
```
Comment (neosbob) [inline README.md:205]: First extracts "webserver" into the deploy directory, and then it changes directory to "deploy" and builds the shrunken image using the files in directory "deploy".

**Pull request merged**